File: LeetCode/Python3/Problem 118.py
# LeetCode: Problem 118
# Jonathan Kosasih

#Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.
#In Pascal's triangle, each number is the sum of the two numbers directly above it.
#
#Example:
#Input: 5
#Output:
#[
#     [1],
#    [1,1],
#   [1,2,1],
#  [1,3,3,1],
# [1,4,6,4,1]
#]
#------------------------------------------------------------------
class Solution(object):
    # initial solution
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        if(not numRows):
            return []
        if(numRows == 1):
            return [[1]]
        ans = [[1], [1,1]]
        for i in range(2,numRows):
            ans.append([])
            ans[i] = [1 for _ in range(i+1)]
            for j in range(1,len(ans[i])-1):
                ans[i][j] = ans[i-1][j-1] + ans[i-1][j]
        return ans
    
    # more optimized
    def generateAlt(self, numRows):
        pascal = [[1]*(i+1) for i in range(numRows)]
        for i in range(numRows):
            for j in range(1,i):
                pascal[i][j] = pascal[i-1][j-1] + pascal[i-1][j]
        return pascal
    
sol = Solution()
print(str(sol.generate(1)))

File: LeetCode/C++/Problem 118.cpp
// LeetCode: Problem 118
// Jonathan Kosasih
/*
Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;

class Solution {
public:
	vector<vector<int>> generate(int numRows) {
		vector<vector<int>> answer(numRows);
		for (int i = 0; i < numRows; i++) {
			answer[i].resize(i + 1);
			answer[i][0] = answer[i][i] = 1; // fill out sides of triangle
			for (int j = 1; j < i; j++) {
				answer[i][j] = answer[i - 1][j - 1] + answer[i - 1][j];
			}
		}
		return answer;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}